python/mlx/nn/utils.py

```diff
@@ -31,3 +34,53 @@ def wrapped_value_grad_fn(*args, **kwargs):
         return value, grad
 
     return wrapped_value_grad_fn
+
+
+def maybe_checkpoint(fn: Callable):
+    """Transform the passed function ``fn`` to a function that optionally
+    performs gradient checkpointing.
+
+    If the function's first argument is a module (always true when it is a
+    method) then the checkpointing considers the module's parameters.
+
+    To check if checkpointing needs to be applied, the transformed function
+    looks at the ``checkpoint`` keyword argument or the ``checkpoint``
+    attribute in case it is a module call.
+
+    Args:
+        fn (Callable): The function to be checkpointed
+
+    Returns:
+        A function that maybe applies gradient checkpointing when calling
+        ``fn``
+    """
+
+    @wraps(fn)
+    def checkpointable_fn(*args, **kwargs):
+        checkpoint = False
+        module = None
+        if isinstance(args[0], Module):
+            module = args[0]
+            if hasattr(module, "checkpoint"):
+                checkpoint = module.checkpoint
```

Comment:
This is intended to be a property you can add to any module right?

So usage:

```python
lin = nn.Linear(10, 10)
lin.checkpoint = True
```

Reply:
Yeah. Basically, one can control checkpointing by either setting an instance property or passing the flag as a keyword argument.